space_packet_parser/definitions.py

```diff
@@ -385,269 +359,326 @@ def _get_container_base_container(
+            if buffer_read_size_bytes is None:
+                # Default to 4096 bytes from a socket
+                buffer_read_size_bytes = 4096
+
+        # ========
+        # Packet loop. Each iteration of this loop yields a ParsedPacket object
+        # ========
+        start_time = time.time_ns()
+        n_bytes_parsed = 0  # Keep track of how many bytes we have parsed
+        n_packets_parsed = 0  # Keep track of how many packets we have parsed
+        read_buffer = b""  # Empty bytes object to start
+        skip_header_bytes = skip_header_bits // 8  # Internally keep track of bytes
+        current_pos = 0  # Keep track of where we are in the buffer
+        while True:
+            if total_length_bytes and n_bytes_parsed == total_length_bytes:
+                break  # Exit if we know the length and we've reached it
+
+            if show_progress is True:
+                self._print_progress(current_bytes=n_bytes_parsed, total_bytes=total_length_bytes,
+                                     start_time_ns=start_time, current_packets=n_packets_parsed)
+
+            if current_pos > 20_000_000:
+                # Only trim the buffer after 20 MB read to prevent modifying
+                # the bitstream and trimming after every packet
+                read_buffer = read_buffer[current_pos:]
+                current_pos = 0
+
+            # Fill buffer enough to parse a header
+            while len(read_buffer) - current_pos < skip_header_bytes + CCSDS_HEADER_LENGTH_BYTES:
+                result = read_bytes_from_source(binary_data, read_size_bytes=buffer_read_size_bytes)
+                if not result:  # If there is verifiably no more data to add, break
+                    break
+                read_buffer += result
+            # Skip the header bytes
+            current_pos += skip_header_bytes
+            header_bytes = read_buffer[current_pos:current_pos + CCSDS_HEADER_LENGTH_BYTES]
+            header = self._parse_header(header_bytes)
+
+            # per the CCSDS spec
+            # 4.1.3.5.3 The length count C shall be expressed as:
+            #   C = (Total Number of Octets in the Packet Data Field) – 1
+            n_bytes_data = header['PKT_LEN'].raw_value + 1
+            n_bytes_packet = CCSDS_HEADER_LENGTH_BYTES + n_bytes_data
+
+            # Consider it a counted packet once we've parsed the header
+            # and update the number of bits parsed
+            n_packets_parsed += 1
+            n_bytes_parsed += skip_header_bytes + n_bytes_packet
+            if ccsds_headers_only is True:
+                # update the current position to the end of the packet data
+                current_pos += n_bytes_packet
+                p = parseables.CCSDSPacket(raw_data=read_buffer[current_pos - n_bytes_packet:current_pos], **header)
+                yield p
+                continue
+
+            # Based on PKT_LEN fill buffer enough to read a full packet
```

Comment:
Thinking about the logic here, should we actually move this while-loop above the header yield/continue? In that yield statement above we are skipping forward `n_bytes_packet`, but we don't know we have that many bytes available in the buffer yet. So we could be advancing past the end of the read_buffer without realizing it. If we move this read above that I think we would fix that issue and shouldn't really slow anything down because we'd have to read that data anyways...

Reply:
Good catch!